### PR TITLE
Limit joint windup for single IK solutions

### PR DESCRIPTION
**Pull request by henningkayser**
This is the first part on limiting joint windup for single IK solutions by setting a "soft" joint limit that still allows exceeding it intentionally. It's based on setting a joint value modifier in range [0,1] that is multiplied with the IK seed state to "pull" desired joints towards the center. This works best in combination with minimal-distance IK solvers (like currently LMA). To eliminate outliers completely the consistency limit of specified joints is set to the lower value PI.

This method allows to preserve the full freedom of the joint since only the seed state is modified.
This allows for consecutive IK solutions to still move the full way if the steps are sufficiently small (i.e. for cartesian planning). For example a value of 0.5 would allow the joint value to converge to 2*PI but only if sequential IK calls produce solutions within closest proximity. The closer the the value is to the limit the lower the distance needs to be to still move towards it. In practice this means that the actually reachable limit is considerably below that one. Lower fractions let the value converge to lower limits (I could make a table for that). A good value to limit wrist windup would be 0.3 or maybe even lower (~1.4*PI limit).

The modifiers are stored in a map `ik_seed_state_fractions_` and can only accessed by getter() and setter() functions that also test the validity of the values. The feature is enabled by calling `robot.limitJointWindup(true)` and is active for all `getJointSolution()` calls including waypoint conversions.

The test shows how this is applied to limit wrist joint windup by setting the corresponding fraction to 0.3.
I suggest specifying desired values either inside the `kinematics.yaml` or inside a new `moveit_simple.yaml` since I think this should be needed anyway.

This solution is meant to be used for all IK calls but does not support end effector symmetries.
For this a separate function will be needed anyway just as for separate pick/place state optimization.
I'm planning to do follow-up PRs for that so this one doesn't get bloated.

### CODE REVIEW

**Comment by mlautman** [comment]
Let's point this at the upstream repo with a note that it should target a devel branch instead

File: moveit_simple/CMakeLists.txt [hunk 1]
```diff
@@ -87,6 +89,10 @@ if(CATKIN_ENABLE_TESTING)
   catkin_add_gtest(moveit_simple_utest ${MOVEIT_SIMPLE_TEST_SRC_FILES})
   target_link_libraries(moveit_simple_utest ${PROJECT_NAME} ${catkin_LIBRARIES} ${Boost_LIBRARIES})
 
+  set(TOOL_WINDUP_SRC_FILES test/tool_windup_utest.cpp test/tool_windup.cpp)
```
Comment (davetcoleman) [inline moveit_simple/CMakeLists.txt:105]: Add comment describing what this does

File: moveit_simple/include/moveit_simple/robot.h [hunk 2]
```diff
@@ -433,6 +433,32 @@ class Robot
...
+  void limitJointWindup(bool enabled);
+
+  /**
+   * @brief setIKSeedStateFractions - Specify joint value modifiers that should be applied in IK
+   * calls if joint windup limitation is enabled. On how to enable and use this feature see the function
+   * limitJointWindup() above.
+   * @param ik_seed_state_fractions - A map from joint ids to fraction values in range [0,1]
+   * @return true on success, false if invalid entries are found (invalid joint id, value out of range)
+   */
+  bool setIKSeedStateFractions(const std::map<size_t, double>& ik_seed_state_fractions);
```
Comment (davetcoleman) [inline moveit_simple/include/moveit_simple/robot.h:461]: These parameters seem too low level for the intended API Shawn wants for the Robot class. Seems to me settings for this should be set in a yaml and just assumed correct from the Robot class.

File: moveit_simple/include/moveit_simple/robot.h [hunk 3]
```diff
@@ -557,10 +583,11 @@ class Robot
   virtual std::vector<double> getJointState(void) const;
 
   bool getIK(const Eigen::Affine3d pose, const std::vector<double> &seed,
-    std::vector<double> &joint_point, double timeout = 1, unsigned int attempts = 1) const;
+    std::vector<double> &joint_point, double timeout = 1, unsigned int attempts = 1,
+    bool limit_joint_windup = false) const;
 
   bool getIK(const Eigen::Affine3d pose, std::vector<double> &joint_point,
-    double timeout = 1, unsigned int attempts = 1) const;
+    double timeout = 1, unsigned int attempts = 1, bool limit_joint_windup = false) const;
```
Comment (davetcoleman) [inline moveit_simple/include/moveit_simple/robot.h]: ``limit_joint_windup`` - i like this level of API visibility for top level users

File: moveit_simple/src/robot.cpp [hunk 4]
```diff
@@ -417,42 +417,33 @@ bool Robot::getJointSolution(const Eigen::Affine3d& pose, double timeout, const
     local_seed = getJointState();
   }
 
-  return getIK(pose, local_seed, joint_point, timeout);
+  return getIK(pose, local_seed, joint_point, timeout, 1, limit_joint_windup_);
```
Comment (davetcoleman) [inline moveit_simple/src/robot.cpp]:
```suggestion
  const unsigned int attempts = 1;
  return getIK(pose, local_seed, joint_point, timeout, attempts, limit_joint_windup_);
```
improves documentation

File: moveit_simple/src/robot.cpp [hunk 5]
```diff
@@ -1502,17 +1493,35 @@ bool Robot::getFK(const std::vector<double>& joint_point, Eigen::Affine3d& pose)
...
+  std::vector<double> local_seed(seed);
+  // If joint windup should be limited, the seed state is 'pulled' towards 0 for the specified joints
+  if (limit_joint_windup)
+  {
+    for (const std::pair<size_t, double> seed_state_fraction : ik_seed_state_fractions_)
+    {
+      size_t joint = seed_state_fraction.first;
+      local_seed[joint] = local_seed[joint] * seed_state_fraction.second;
+    }
+  }
```
Comment (davetcoleman) [inline moveit_simple/src/robot.cpp:1246]: make this a private helper function ``reduceJointWindup`` or ``biasSeedState``

File: moveit_simple/src/robot.cpp [hunk 6]
```diff
@@ -1526,6 +1535,36 @@ bool Robot::getIK(const Eigen::Affine3d pose, std::vector<double>& joint_point,
   return false;
 }
 
+void Robot::limitJointWindup(bool enabled)
+{
+  limit_joint_windup_ = enabled;
+}
```
Comment (davetcoleman) [inline moveit_simple/src/robot.cpp:1279]: i don't like how ``getIK`` has a ``limit_joint_windup`` but this public function also exist. should be one or the other...

File: moveit_simple/src/robot.cpp [hunk 7]
```diff
@@ -1526,6 +1535,36 @@ bool Robot::getIK(const Eigen::Affine3d pose, std::vector<double>& joint_point,
...
+
+bool Robot::setIKSeedStateFractions(const std::map<size_t, double>& ik_seed_state_fractions)
+{
+  size_t joint_num = joint_group_->getActiveJointModels().size();
```
Comment (davetcoleman) [inline moveit_simple/src/robot.cpp:1288]: const

File: moveit_simple/test/tool_windup.cpp [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+ * Objects that can be directly used inside the test
+ - robot pointer for moveit_simple::OnlineRobot
+*/
+class UserRobotTest : public ::testing::Test
+{
+protected:
+  std::unique_ptr<moveit_simple::OnlineRobot> robot;
+  virtual void SetUp()
+  {
+  robot = std::unique_ptr<moveit_simple::OnlineRobot> (new moveit_simple::OnlineRobot
```
Comment (davetcoleman) [inline moveit_simple/test/tool_windup.cpp]: indent (clang format)

File: moveit_simple/test/tool_windup.cpp [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+ - robot pointer for DeveloperRobot
+*/
+
+class DeveloperRobotTest : public ::testing::Test
+{
+protected:
+  std::unique_ptr<DeveloperRobot> robot2;
+  virtual void SetUp()
+  {
+  robot2 = std::unique_ptr<DeveloperRobot> (new DeveloperRobot
```
Comment (davetcoleman) [inline moveit_simple/test/tool_windup.cpp]: indent

File: moveit_simple/test/tool_windup.cpp [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+/**
+ * @brief DeveloperRobotTest is a fixture for testing protected methods.
+ * Objects that can be directly used inside the test
+ - robot pointer for DeveloperRobot
+*/
+
+class DeveloperRobotTest : public ::testing::Test
+{
+protected:
+  std::unique_ptr<DeveloperRobot> robot2;
```
Comment (davetcoleman) [inline moveit_simple/test/tool_windup.cpp]:
```suggestion
  std::unique_ptr<DeveloperRobot> robot_;
```
why 2?

File: moveit_simple/test/tool_windup_utest.cpp [hunk 11]
```diff
@@ -0,0 +1,34 @@
+/*
+ * Software License Agreement (Apache License)
+ *
+ * Copyright (c) 2019 Plus One Robotics
```
Comment (davetcoleman) [inline moveit_simple/test/tool_windup_utest.cpp]: our license

File: moveit_simple/test/tool_windup.cpp [hunk 12]
```diff
@@ -0,0 +1,149 @@
+/*
+ * Software License Agreement (Apache License)
+ *
+ * Copyright (c) 2019 Plus One Robotics
```
Comment (davetcoleman) [inline moveit_simple/test/tool_windup.cpp]: our license

File: moveit_simple/CMakeLists.txt (hunk 1 above)
Comment (henningkayser) [reply moveit_simple/CMakeLists.txt:105]: Inside the test source or directly inside the CMakeLists?

File: moveit_simple/include/moveit_simple/robot.h (hunk 2 above)
Comment (henningkayser) [reply moveit_simple/include/moveit_simple/robot.h:461]: Sure, I thought the same thing. I will add a demo config file to the test cases.

File: moveit_simple/include/moveit_simple/robot.h (hunk 3 above)
Comment (henningkayser) [reply moveit_simple/include/moveit_simple/robot.h]: Yes, `getIK()` is developer level so I made it optional here, but `getJointSolution()` is user level and the value passed to `limitJointWindup()` is used.

File: moveit_simple/src/robot.cpp (hunk 4 above)
Comment (henningkayser) [reply moveit_simple/src/robot.cpp]: Probably makes sense. The `attempts` parameter will be removed when we're switching to the master branch, though.

File: moveit_simple/src/robot.cpp (hunk 6 above)
Comment (henningkayser) [reply moveit_simple/src/robot.cpp:1279]: See https://github.com/PickNikRobotics/moveit_simple/pull/3#discussion_r279870827.
It's a matter of what user level is accessing the function. A normal user only has access to the public function `getJointSolutions()` which uses `limit_joint_windup_` and calls `getIK()`. `getIK()` is protected and is free to be used with and without joint windup reduction, i.e. when sampling with symmetries (later PR).

File: moveit_simple/test/tool_windup.cpp (hunk 10 above)
Comment (henningkayser) [reply moveit_simple/test/tool_windup.cpp]: That's taken from other test sources where there is a `UserRobot` "robot" and a `DeveloperRobot` "robot2"

**Pull request closed without merging**